File: mergeKsorted.py
#TC - O(NlogK)
#SC - O(N)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        for i,l in enumerate(lists):
            if l:
                heappush(heap,(l.val,i))
        curr = dummy = ListNode(0)
        while heap:
            val, i = heappop(heap)
            curr.next = ListNode(val)
            if lists[i].next:
                heappush(heap,(lists[i].next.val,i))
                lists[i] = lists[i].next
            curr = curr.next
        return dummy.next

            

        

File: mergeKSorted.cpp

//TIMEO(N log N)
//SPACE(O(n))


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    struct CompareNodes {
    bool operator()(const ListNode* a, const ListNode* b) {
        return a->val > b->val; // Min heap, smallest value has highest priority
    }
    };
    priority_queue<ListNode* , vector<ListNode*>, CompareNodes> pq;
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        for(ListNode* head:lists){
            if(head != nullptr)
            {
                pq.push(head);

            }
        }
        ListNode* dummy = new ListNode(-1);
        ListNode* curr = dummy;
        while(!pq.empty()){
            ListNode* currmin =pq.top();
            pq.pop();
            
            curr->next = currmin;
            curr = curr->next;

            

            if(currmin->next != nullptr){
                pq.push(currmin->next);
            }





        }

        return dummy->next;
        
        
    }
};